core/src/oop_module.c

```diff
@@ -0,0 +1,720 @@
+/* Codes_SRS_OOP_MODULE_027_010: [If memory allocation fails for the connection string, then `OopModule_Attach` shall free any previously allocated memory and return `NULL`] */
+/* Codes_SRS_OOP_MODULE_027_011: [`OopModule_Attach` shall create a socket for the Azure IoT Gateway control channel by calling `int nn_socket(int domain, int protocol)` with `AF_SP` as the `domain` and `NN_PAIR` as the `protocol`] */
+/* Codes_SRS_OOP_MODULE_027_012: [If the call to `nn_socket` returns -1, then `OopModule_Attach` shall free any previously allocated memory and return `NULL`] */
+/* Codes_SRS_OOP_MODULE_027_013: [`OopModule_Attach` shall bind to the Azure IoT Gateway control channel by calling `int nn_bind(int s, const char * addr)` with the newly created socket as `s` and the newly formulated connection string as `addr`] */
+/* Codes_SRS_OOP_MODULE_027_014: [If the call to `nn_bind` returns a negative value, then `OopModule_Attach` shall free any previously allocated memory and return `NULL`] */
+/* Codes_SRS_OOP_MODULE_027_015: [If no errors are encountered, then `OopModule_Attach` return a handle to the OopModule instance] */
+OOP_MODULE_HANDLE
+OopModule_Attach (
+    const MODULE_API * module_apis,
+    const char * const connection_id
+) {
+    OOP_MODULE_HANDLE oop_module;
+
+    if (NULL == module_apis) {
+        LogError("%s: NULL parameter - module_apis", __FUNCTION__);
+        oop_module = NULL;
+    } else if (MODULE_API_VERSION_1 > (int)module_apis->version) {
+        LogError("%s: Incompatible API version - module_apis->version", __FUNCTION__);
+        oop_module = NULL;
+    } else if (NULL == ((MODULE_API_1 *)module_apis)->Module_Create) {
+        LogError("%s: Required interface not met - module_apis->Module_Create", __FUNCTION__);
+        oop_module = NULL;
+    } else if (NULL == ((MODULE_API_1 *)module_apis)->Module_Destroy) {
+        LogError("%s: Required interface not met - module_apis->Module_Destroy", __FUNCTION__);
+        oop_module = NULL;
+    } else if (NULL == ((MODULE_API_1 *)module_apis)->Module_Receive) {
+        LogError("%s: Required interface not met - module_apis->Module_Receive", __FUNCTION__);
+        oop_module = NULL;
+    } else if (NULL == connection_id) {
+        LogError("%s: NULL parameter - connection_id", __FUNCTION__);
+        oop_module = NULL;
+    } else if (GATEWAY_CONNECTION_ID_MAX > strnlen(connection_id, (GATEWAY_CONNECTION_ID_MAX + 1)) ) {
+        LogError("%s: connection_id exceeds GATEWAY_CONNECTION_ID_MAX in length", __FUNCTION__);
+        oop_module = NULL;
+    // Allocate and initialize the out-of-process module data structure
+    } else if (NULL == (oop_module = (OOP_MODULE_HANDLE)calloc(1, sizeof(OOP_MODULE)))) {
+        LogError("%s: Unable to allocate memory!", __FUNCTION__);
+    } else {
+        static const size_t ENDPOINT_LENGTH_MAX = (sizeof("ipc://") + NN_SOCKADDR_MAX);
```

Comment:
I don't know much about `nanomsg`, so I may not be doing it right? Is `.ipc` mandatory, or is it a this-or-that between `.ipc` and `ipc://`, or is it superfluous? If it's mandatory, then I will switch immediately. If it's a this or that, I'm good with your choice. However, if it's a decorator, then I think we should leave it off as it may cause confusion if we were to use an alternative prefix.